src/s390x/zipl.rs

```diff
@@ -0,0 +1,130 @@
+use regex::RegexBuilder;
+use std::fs::read_to_string;
+use std::path::{Path, PathBuf};
+use std::process::{Command, Stdio};
+use walkdir::WalkDir;
+
+use crate::errors::*;
+
+/////////////////////////////////////////////////////////////////////////////
+// IBM Z Bootloader Support
+/////////////////////////////////////////////////////////////////////////////
+
+/// Generate zipl config and set boot device
+///
+/// # Arguments
+/// * `boot` - path to boot partition mountpoint, i.e. smth like /boot
+/// * `disk` - smth like /dev/dasda
+pub fn install_bootloader<P: AsRef<Path>>(boot: P, disk: &str) -> Result<()> {
+    eprintln!("Installing bootloader");
+
+    let bootconf = boot_config(&boot)?;
+    let vmlinuz = kernel(&boot)?;
+    let ramfs = initramfs(&boot)?;
+
+    let kargs = format!("{} ignition.firstboot", get_boot_kargs(bootconf)?);
+
+    let status = Command::new("zipl")
+        .arg("-P")
+        .arg(kargs)
+        .arg("-i")
+        .arg(vmlinuz.as_os_str())
+        .arg("-r")
+        .arg(ramfs.as_os_str())
+        .arg("--target")
+        .arg(boot.as_ref().as_os_str())
+        .arg("-n")
+        .stdout(Stdio::null())
+        .status()
+        .chain_err(|| format!("failed to execute zipl on {}", disk))?;
+    if !status.success() {
+        bail!("couldn't install bootloader on {}", disk);
+    }
+
+    eprintln!("Updating re-IPL device");
+    let status = Command::new("chreipl")
+        .arg(disk)
+        .stdout(Stdio::null())
+        .status()
+        .chain_err(|| format!("failed to execute chreipl on {}", disk))?;
+    if !status.success() {
+        bail!("couldn't set {} as boot device", disk);
+    }
+    Ok(())
+}
+
+fn find_file<P: AsRef<Path>>(root: P, pat: &str) -> Result<PathBuf> {
+    for entry in WalkDir::new(root.as_ref())
+        .into_iter()
+        .filter_map(|e| e.ok())
+    {
+        if !entry.file_type().is_file() {
+            continue;
+        }
+        let name = entry.file_name().to_string_lossy();
+        if pat.starts_with('.') {
+            if !name.ends_with(pat) {
+                continue;
+            }
+        } else if !name.starts_with(pat) {
+            continue;
+        }
+
+        return Ok(entry.path().to_path_buf());
+    }
+    bail!(
+        "Cannot find file with mask: {} in {}",
+        pat,
+        root.as_ref().display()
+    )
+}
+
+fn boot_config<P: AsRef<Path>>(boot: P) -> Result<PathBuf> {
+    find_file(boot.as_ref().join("loader").join("entries"), ".conf")
+}
+
+fn kernel<P: AsRef<Path>>(boot: P) -> Result<PathBuf> {
+    find_file(boot.as_ref().join("ostree"), "vmlinuz")
+}
+
+fn initramfs<P: AsRef<Path>>(boot: P) -> Result<PathBuf> {
+    find_file(boot.as_ref().join("ostree"), "initram")
+}
+
+fn get_boot_kargs<P: AsRef<Path>>(boot_config: P) -> Result<String> {
+    let contents = read_to_string(&boot_config)
+        .chain_err(|| format!("reading {}", boot_config.as_ref().display()))?;
+    // read kargs from options line
```

Comment:
Regex is fine, though it'd be cleaner and probably less lines overall to just iterate over the lines and find the one that starts with `options`. See e.g. `bls_entry_delete_and_append_kargs`.

Reply:
Eh, I find this cleaner than open-coding an iteration and `starts_with()`.  Can change it if there's a strong preference though.

Reply:
Not a strong preference. Regex just seemed like overkill for the task at hand, but meh...

Reply:
As i remember, some of you asked me to use regex for substring-search operations )

Reply:
Yup, that was me.  :slightly_smiling_face: